File: spec/services/auth_service_spec.rb
# frozen_string_literal: true

require 'spec_helper'

RSpec.describe SinatraBoilerplate::Modules::Auth::Service do
  let(:service) { described_class.new }

  describe '#register' do
    it 'creates a user and returns a token pair' do
      result = service.register(
        name: 'New User',
        email: 'svc@example.com',
        password: 'password123'
      )

      expect(result[:access_token]).to be_a(String)
      expect(result[:refresh_token]).to be_a(String)
      expect(result[:user].email).to eq('svc@example.com')
    end

    it 'raises on duplicate email' do
      create(:user, email: 'dup@example.com')
      expect do
        service.register(name: 'X', email: 'dup@example.com', password: 'password123')
      end.to raise_error(SinatraBoilerplate::Errors::ConflictError)
    end
  end

  describe '#login' do
    it 'returns a token pair on valid credentials' do
      create(:user, email: 'login@example.com', password: 'login12345')
      result = service.login(email: 'login@example.com', password: 'login12345')
      expect(result[:access_token]).to be_a(String)
    end

    it 'raises AuthenticationError on wrong password' do
      create(:user, email: 'login@example.com', password: 'login12345')
      expect do
        service.login(email: 'login@example.com', password: 'wrong')
      end.to raise_error(SinatraBoilerplate::Errors::AuthenticationError)
    end
  end

  describe '#refresh' do
    it 'rotates the refresh token' do
      create(:user, email: 'rot@example.com', password: 'rot1234567')
      login = service.login(email: 'rot@example.com', password: 'rot1234567')
      old = login[:refresh_token]

      result = service.refresh(old)
      expect(result[:refresh_token]).not_to eq(old)
    end
  end

  describe '#logout_all' do
    it 'revokes all tokens for a user' do
      user = create(:user, email: 'all@example.com')
      service.logout_all(user)
      expect(SinatraBoilerplate::Modules::Auth::RefreshToken.where(user_id: user.id).count).to eq(0)
    end
  end
end

File: test/unit/auth.service.spec.ts
import { describe, it, expect } from 'vitest';
import * as argon2 from 'argon2';
import { AuthService } from '@/modules/auth/auth.service';
import { type UsersService } from '@/modules/users/services/users.service';
import { UsersRepository } from '@/modules/users/repositories/users.repository';
import { type JwtService } from '@nestjs/jwt';
import { type ConfigService } from '@nestjs/config';
import { type PrismaService } from '@/infrastructure/database/prisma.service';
import type { LoginInput, RegisterInput } from '@/modules/auth/schemas';

class FakeUsersService {
  async create(payload: RegisterInput) {
    return { id: 'u1', email: payload.email.toLowerCase(), name: payload.name, role: 'USER' as const, status: 'ACTIVE' as const, createdAt: new Date().toISOString(), updatedAt: new Date().toISOString() };
  }
  async findRawByEmail(email: string) {
    if (email !== 'alice@example.com') return null;
    return {
      id: 'u1',
      email: 'alice@example.com',
      name: 'Alice',
      password: await argon2.hash('Strong1Pass'),
      role: 'USER' as const,
      status: 'ACTIVE' as const,
      createdAt: new Date(),
      updatedAt: new Date(),
      deletedAt: null,
    };
  }
  async findById() {
    throw new Error('not used');
  }
}

class FakeJwtService {
  async signAsync(payload: object, options?: { expiresIn?: string | number }) {
    return `token.${JSON.stringify({ ...payload, options })}.sig`;
  }
}

class FakeConfigService {
  get(key: string) {
    const values: Record<string, string> = {
      'auth.jwtSecret': 'a'.repeat(32),
      'auth.jwtAccessExpiresIn': '15m',
      'auth.jwtRefreshExpiresIn': '7d',
    };
    return values[key];
  }
}

class FakePrismaService {
  refreshToken = {
    async create() {
      return {} as never;
    },
    async findUnique() {
      return null;
    },
    async update() {
      return {} as never;
    },
    async updateMany() {
      return { count: 0 };
    },
  };
  async $connect() {}
  async $disconnect() {}
}

describe('AuthService', () => {
  const makeService = () => {
    const users = new FakeUsersService();
    const fakePrisma = new FakePrismaService();
    const fakeJwt = new FakeJwtService();
    const fakeConfig = new FakeConfigService();
    return new AuthService(
      users as unknown as UsersService,
      fakeJwt as unknown as JwtService,
      fakeConfig as unknown as ConfigService,
      fakePrisma as unknown as PrismaService,
    );
  };

  it('registers a new account and returns tokens', async () => {
    const svc = makeService();
    const result = await svc.register({ name: 'Alice', email: 'alice@example.com', password: 'Strong1Pass' });
    expect(result.accessToken).toBeTruthy();
    expect(result.refreshToken).toBeTruthy();
    expect(result.tokenType).toBe('Bearer');
    expect(result.expiresIn).toBeGreaterThan(0);
  });

  it('logs a user in with valid credentials', async () => {
    const svc = makeService();
    const loginInput: LoginInput = { email: 'alice@example.com', password: 'Strong1Pass' };
    const result = await svc.login(loginInput);
    expect(result.accessToken).toBeTruthy();
  });

  it('rejects login with wrong password', async () => {
    const svc = makeService();
    await expect(svc.login({ email: 'alice@example.com', password: 'WrongPass1' })).rejects.toThrow();
  });

  it('rejects login for unknown user with a generic error', async () => {
    const svc = makeService();
    await expect(svc.login({ email: 'ghost@example.com', password: 'whatever1' })).rejects.toThrow();
  });
});

void UsersRepository;
